### PR TITLE
compiler: implement `@branchHint`

### PR DESCRIPTION
**Pull request by mlugg**
See commit messages for details.

### CODE REVIEW

**Comment by mlugg** [comment]
@Rexicon226 confirmed that this seems to be getting the desired codegen.

File: test/behavior/builtin_functions_returning_void_or_noreturn.zig [hunk 1]
```diff
@@ -22,7 +22,7 @@ test {
     try testing.expectEqual(noreturn, @TypeOf(if (true) @panic("") else {}));
     try testing.expectEqual({}, @prefetch(&val, .{}));
     try testing.expectEqual({}, @setAlignStack(16));
-    try testing.expectEqual({}, @setCold(true));
+    try testing.expectEqual({}, @branchHint(.cold));
```
Comment (andrewrk) [inline test/behavior/builtin_functions_returning_void_or_noreturn.zig:25]: This doesn't trip `'@branchHint' must appear as the first statement in a function or conditional branch`?

**Review by andrewrk — APPROVED** [review]
It will be interesting to take a performance data point to find out if weighting branches against errors had any effect on compiler performance.

File: test/behavior/builtin_functions_returning_void_or_noreturn.zig (hunk 1 above)
Comment (mlugg) [reply test/behavior/builtin_functions_returning_void_or_noreturn.zig:25]: huh, yeah, that's... wrong. good spot

**Comment by mlugg** [comment]
Superceded by #21214.

**Pull request closed without merging**